File: app/src/main/java/com/elementary/tasks/core/utils/Language.java
package com.elementary.tasks.core.utils;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.DisplayMetrics;

import com.elementary.tasks.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class Language {
    public static final String ENGLISH = "en";
    public static final String FRENCH = "fr";
    public static final String GERMAN = "de";
    public static final String ITALIAN = "it";
    public static final String JAPANESE = "ja";
    public static final String KOREAN = "ko";
    public static final String POLISH = "pl";
    public static final String RUSSIAN = "ru";
    public static final String SPANISH = "es";
    public static final String UKRAINIAN = "uk";

    public static final String EN = "en-US";
    public static final String RU = "ru-RU";
    public static final String UK = "uk-UA";

    public static Context onAttach(Context context) {
        return setLocale(context, getScreenLanguage(Prefs.getInstance(context).getAppLanguage()));
    }

    public static Context setLocale(Context context, Locale locale) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return updateResources(context, locale);
        }
        return updateResourcesLegacy(context, locale);
    }

    @TargetApi(Build.VERSION_CODES.N)
    private static Context updateResources(Context context, Locale locale) {
        Locale.setDefault(locale);
        Configuration configuration = context.getResources().getConfiguration();
        configuration.setLocale(locale);
        configuration.setLayoutDirection(locale);
        return context.createConfigurationContext(configuration);
    }

    @SuppressWarnings("deprecation")
    private static Context updateResourcesLegacy(Context context, Locale locale) {
        Locale.setDefault(locale);
        Resources resources = context.getResources();
        Configuration configuration = resources.getConfiguration();
        configuration.locale = locale;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            configuration.setLayoutDirection(locale);
        }
        resources.updateConfiguration(configuration, resources.getDisplayMetrics());
        return context;
    }

    @NonNull
    public static String getLocalized(Context context, int id) {
        if (Module.isJellyMR1()) {
            Configuration configuration = new Configuration(context.getResources().getConfiguration());
            configuration.setLocale(new Locale(Language.getTextLanguage(Prefs.getInstance(context).getVoiceLocale())));
            return context.createConfigurationContext(configuration).getResources().getString(id);
        } else {
            Resources standardResources = context.getResources();
            AssetManager assets = standardResources.getAssets();
            DisplayMetrics metrics = standardResources.getDisplayMetrics();
            Configuration config = new Configuration(standardResources.getConfiguration());
            config.locale = new Locale(Language.getTextLanguage(Prefs.getInstance(context).getVoiceLocale()));
            Resources defaultResources = new Resources(assets, metrics, config);
            return defaultResources.getString(id);
        }
    }

    @NonNull
    public static List<String> getLanguages(Context context) {
        List<String> locales = new ArrayList<>();
        locales.add(context.getString(R.string.english) + " (" + EN + ")");
        locales.add(context.getString(R.string.russian) + " (" + RU + ")");
        locales.add(context.getString(R.string.ukrainian) + " (" + UK + ")");
        return locales;
    }

    @NonNull
    public static Locale getScreenLanguage(int code) {
        switch (code) {
            case 0:
                return Locale.getDefault();
            case 1:
                return Locale.ENGLISH;
            case 2:
                return Locale.GERMAN;
            case 3:
                return new Locale("es", "");
            case 4:
                return Locale.FRENCH;
            case 5:
                return Locale.ITALIAN;
            case 6:
                return new Locale("pt", "");
            case 7:
                return new Locale("pl", "");
            case 8:
                return new Locale("cs", "");
            case 9:
                return new Locale("ro", "");
            case 10:
                return new Locale("tr", "");
            case 11:
                return new Locale("uk", "");
            case 12:
                return new Locale("ru", "");
            case 13:
                return Locale.JAPANESE;
            case 14:
                return Locale.CHINESE;
            case 15:
                return new Locale("hi", "");
            default:
                return Locale.getDefault();
        }
    }

    @NonNull
    public static String getTextLanguage(int code) {
        switch (code) {
            case 0:
                return ENGLISH;
            case 1:
                return RUSSIAN;
            case 2:
                return UKRAINIAN;
            default:
                return ENGLISH;
        }
    }

    @NonNull
    public static String getLanguage(int code) {
        switch (code) {
            case 0:
                return EN;
            case 1:
                return RU;
            case 2:
                return UK;
            default:
                return EN;
        }
    }

    @NonNull
    public static String getLocaleByPosition(int position) {
        String locale = Language.ENGLISH;
        if (position == 0) locale = Language.ENGLISH;
        if (position == 1) locale = Language.FRENCH;
        if (position == 2) locale = Language.GERMAN;
        if (position == 3) locale = Language.ITALIAN;
        if (position == 4) locale = Language.JAPANESE;
        if (position == 5) locale = Language.KOREAN;
        if (position == 6) locale = Language.POLISH;
        if (position == 7) locale = Language.RUSSIAN;
        if (position == 8) locale = Language.SPANISH;
        if (position == 9 && Module.isJellyMR2()) locale = Language.UKRAINIAN;
        return locale;
    }

    public static int getLocalePosition(@Nullable String locale) {
        if (locale == null) {
            return 0;
        }
        int mItemSelect = 0;
        if (locale.matches(Language.ENGLISH)) {
            mItemSelect = 0;
        } else if (locale.matches(Language.FRENCH)) {
            mItemSelect = 1;
        } else if (locale.matches(Language.GERMAN)) {
            mItemSelect = 2;
        } else if (locale.matches(Language.ITALIAN)) {
            mItemSelect = 3;
        } else if (locale.matches(Language.JAPANESE)) {
            mItemSelect = 4;
        } else if (locale.matches(Language.KOREAN)) {
            mItemSelect = 5;
        } else if (locale.matches(Language.POLISH)) {
            mItemSelect = 6;
        } else if (locale.matches(Language.RUSSIAN)) {
            mItemSelect = 7;
        } else if (locale.matches(Language.SPANISH)) {
            mItemSelect = 8;
        } else if (locale.matches(Language.UKRAINIAN) && Module.isJellyMR2()) {
            mItemSelect = 9;
        }
        return mItemSelect;
    }

    @NonNull
    public static String getScreenLocaleName(Context context) {
        return context.getResources().getStringArray(R.array.app_languages)[Prefs.getInstance(context).getAppLanguage()];
    }

    @NonNull
    public static List<String> getScreenLocaleNames(Context context) {
        ArrayList<String> names = new ArrayList<>();
        Collections.addAll(names, context.getResources().getStringArray(R.array.app_languages));
        return names;
    }

    @NonNull
    public static List<String> getLocaleNames(Context mContext) {
        ArrayList<String> names = new ArrayList<>();
        names.add(mContext.getString(R.string.english));
        names.add(mContext.getString(R.string.french));
        names.add(mContext.getString(R.string.german));
        names.add(mContext.getString(R.string.italian));
        names.add(mContext.getString(R.string.japanese));
        names.add(mContext.getString(R.string.korean));
        names.add(mContext.getString(R.string.polish));
        names.add(mContext.getString(R.string.russian));
        names.add(mContext.getString(R.string.spanish));
        if (Module.isJellyMR2()) {
            names.add(mContext.getString(R.string.ukrainian));
        }
        return names;
    }

    /**
     * Holder locale for tts.
     *
     * @param context application context.
     * @param isBirth flag for birthdays.
     * @return Locale
     */
    @Nullable
    public Locale getLocale(Context context, boolean isBirth) {
        Locale res = null;
        String locale;
        if (isBirth) {
            locale = Prefs.getInstance(context).getBirthdayTtsLocale();
        } else {
            locale = Prefs.getInstance(context).getTtsLocale();
        }
        if (locale == null) {
            return Locale.ENGLISH;
        }
        switch (locale) {
            case ENGLISH:
                res = Locale.ENGLISH;
                break;
            case FRENCH:
                res = Locale.FRENCH;
                break;
            case GERMAN:
                res = Locale.GERMAN;
                break;
            case JAPANESE:
                res = Locale.JAPANESE;
                break;
            case ITALIAN:
                res = Locale.ITALIAN;
                break;
            case KOREAN:
                res = Locale.KOREAN;
                break;
            case POLISH:
                res = new Locale("pl", "");
                break;
            case RUSSIAN:
                res = new Locale("ru", "");
                break;
            case SPANISH:
                res = new Locale("es", "");
                break;
            case UKRAINIAN:
                res = new Locale("uk", "");
                break;
        }
        return res;
    }
}

File: app/src/main/java/com/elementary/tasks/core/utils/Language.kt
package com.elementary.tasks.core.utils

import android.content.Context
import android.content.res.Configuration
import com.elementary.tasks.R
import com.elementary.tasks.core.utils.params.Prefs
import com.github.naz013.feature.common.android.TextProvider
import java.util.Locale

class Language(
  private val prefs: Prefs,
  private val context: Context,
  private val textProvider: TextProvider
) {

  fun getCurrentLocale(): String {
    val defLocale = context.resources.configuration.locale
    val locale = runCatching { context.resources.configuration.locales.get(0) }.getOrNull()
      ?: defLocale
    return locale.language
  }

  fun onAttach(context: Context): Context {
    return setLocale(context, getScreenLanguage(prefs.appLanguage)).also {
      textProvider.updateContext(it)
    }
  }

  private fun setLocale(context: Context, locale: Locale): Context {
    return updateResources(context, locale)
  }

  private fun updateResources(context: Context, locale: Locale): Context {
    Locale.setDefault(locale)
    val configuration = context.resources.configuration
    configuration.setLocale(locale)
    configuration.setLayoutDirection(locale)
    return context.createConfigurationContext(configuration)
  }

  fun getConversationLocalizedText(context: Context, id: Int): String {
    val configuration = Configuration(context.resources.configuration)
    configuration.setLocale(Locale(getTextLanguage(prefs.voiceLocale)))
    return context.createConfigurationContext(configuration).resources.getString(id)
  }

  fun getConversationLocalizedContext(): Context {
    val configuration = Configuration(context.resources.configuration)
    configuration.setLocale(Locale(getTextLanguage(prefs.voiceLocale)))
    return context.createConfigurationContext(configuration)
  }

  fun getLanguages(context: Context) = listOf(
    context.getString(R.string.english),
    context.getString(R.string.ukrainian),
    context.getString(R.string.spanish),
    context.getString(R.string.portuguese),
    context.getString(R.string.polish),
    context.getString(R.string.italian)
  )

  fun getTextLanguage(code: Int) = when (code) {
    0 -> ENGLISH
    1 -> UKRAINIAN
    2 -> SPANISH
    3 -> PORTUGUESE
    4 -> POLISH
    5 -> ITALIAN
    else -> ENGLISH
  }

  fun getLanguage(code: Int) = when (code) {
    0 -> EN
    1 -> UK
    2 -> ES
    3 -> PT
    4 -> PL
    5 -> IT
    else -> EN
  }

  fun getVoiceLocale(code: Int): Locale = when (code) {
    0 -> Locale.ENGLISH
    1 -> Locale("uk", "")
    2 -> Locale("es", "")
    3 -> Locale("pt", "")
    4 -> Locale("pl", "")
    5 -> Locale.ITALIAN
    else -> Locale.ENGLISH
  }

  fun getVoiceLanguage(code: Int) = when (code) {
    0 -> com.backdoor.engine.misc.Locale.EN
    1 -> com.backdoor.engine.misc.Locale.UK
    2 -> com.backdoor.engine.misc.Locale.ES
    3 -> com.backdoor.engine.misc.Locale.PT
    4 -> com.backdoor.engine.misc.Locale.PL
    5 -> com.backdoor.engine.misc.Locale.IT
    else -> com.backdoor.engine.misc.Locale.EN
  }

  fun getScreenLocaleName(context: Context): String =
    context.resources.getStringArray(R.array.app_languages)[prefs.appLanguage]

  companion object {
    const val ENGLISH = "en"
    const val FRENCH = "fr"
    const val GERMAN = "de"
    const val ITALIAN = "it"
    const val JAPANESE = "ja"
    const val KOREAN = "ko"
    const val POLISH = "pl"
    const val RUSSIAN = "ru"
    const val SPANISH = "es"
    const val UKRAINIAN = "uk"
    const val PORTUGUESE = "pt"
    const val BULGARIAN = "bg"

    private const val EN = "en-US"
    private const val UK = "uk-UA"
    private const val ES = "es-ES"
    private const val PT = "pt-PT"
    private const val PL = "pl"
    private const val IT = "it"

    fun getScreenLanguage(code: Int): Locale {
      return when (code) {
        0 -> Locale.getDefault()
        1 -> Locale.ENGLISH
        2 -> Locale.GERMAN
        3 -> Locale(SPANISH, "")
        4 -> Locale.FRENCH
        5 -> Locale.ITALIAN
        6 -> Locale(PORTUGUESE, "")
        7 -> Locale(POLISH, "")
        8 -> Locale("cs", "")
        9 -> Locale("ro", "")
        10 -> Locale("tr", "")
        11 -> Locale(UKRAINIAN, "")
        12 -> Locale(RUSSIAN, "")
        13 -> Locale.JAPANESE
        14 -> Locale.CHINESE
        15 -> Locale("hi", "")
        16 -> Locale.KOREAN
        17 -> Locale(BULGARIAN, "")
        else -> Locale.getDefault()
      }
    }
  }
}
